### DISCUSSION TITLE
Option to disable and outsource image transformations

### DISCUSSION THREAD

**Discussion by cemreinanc** [directus/directus#8133] [Feature Requests]
We could have unexpected errors while directus try to process big or corrupt images. And this causes all api&cms to restart. Or at least it causes major performance downgrade for api responses while processing an image in the same instance.

I think it shoud be the best to seperate this image processing layer from the api. While we can already use external products and services to manipulate images for our website frontend, currently its not possible to completely disable this function in api, and force directus app ui to use these external services.

Is it possible to completely disable these image optimiazation&transfomation middleware in directus core api, and use cloudinary or AWS serverless image handler or similiar services as a replacement? I think it requires a custom function in app ui to get source image info, requested width height and other parameters, and return a different cdn url then directus internal apis.

**Comment by joselcvarela** [comment]
Hey,


About external upload, you already can define where you want to upload your files.
Check if this can help you:
https://docs.directus.io/concepts/files/#storage-adapters
https://docs.directus.io/guides/api-config/#multiple-storage-adapters


As for the:
> I think it shoud be the best to seperate this image processing layer from the api.

part, are you suggesting to make a parallel service to handle the uploads? Like https://imgix.com/ ?

**Reply by cemreinanc** [reply]
No its not about the uploads, its about image transformations like generating smaller assets from the original file.
Currently directus api internally process all images using sharp package. And directus app uses this locally transformed versions in the ui. Wheter its stored locally or any other storage like S3.

**Reply by cemreinanc** [reply]
Of course seperating the process also could be better but it also brings another complexities for small deployments.

I'm suggesting "optionally" disabling this internal processing layer, and adding an option use external services like imgix for the admin ui.

**Reply by joselcvarela** [reply]
You can already disable this by putting this `ASSETS_TRANSFORM_IMAGE_MAX_DIMENSION=0` on your `.env` file. But you will not have any previews, only a message saying it cannot show the preview but when you click on the zoom icon you will be able to see the image. Check here: https://github.com/directus/directus/pull/8069

About using external services for the admin UI, I think we should sandbox as much as possible.
I mean, I think we should avoid using an external layer between two layers of ours.
It could be difficult and hard to maintain and also can broke things if the external layer decided to make breaking changes.

**Reply by SalahAdDin** [reply]
> Hey,
> 
> About external upload, you already can define where you want to upload your files. Check if this can help you: https://docs.directus.io/concepts/files/#storage-adapters https://docs.directus.io/guides/api-config/#multiple-storage-adapters
> 
> As for the:
> 
> > I think it shoud be the best to seperate this image processing layer from the api.
> 
> part, are you suggesting to make a parallel service to handle the uploads? Like https://imgix.com/ ?

Documentation links are wrong, sorry.

**Reply by rijkvanzanten** [reply]
The docs will update over time, the search box is the easiest way to find what you're looking for 👍🏻 

<img width="643" alt="CleanShot 2022-08-05 at 15 03 13@2x" src="https://user-images.githubusercontent.com/9141017/183144152-e2cc7ad8-e4ce-4003-bbc6-e586835f5f79.png">

**Comment by rijkvanzanten** [comment]
> We could have unexpected errors while directus try to process big or corrupt images

We already started adding environment variables to control this more fine-grained (https://docs.directus.io/reference/environment-variables/#assets), limiting things like max assets to be transformed at a time, max file size to try transform etc etc. Obviously the goal is to make sure that these sorts of unexpected errors are minimized to a point where it doesn't have to be a worry anymore.

Adding a configuration to simply turn it off altogether is something we could consider/do, but allowing other services to handle that part on the fly might be very difficult, as there's no real specced standards for things like query params, supported features etc. It would have to be another extension type just like storage adapters to integrate with specific third party providers, which adds a lot of technical overhead 🤔

**Reply by benhaynes** [reply]
Can't this be done by setting the project settings to allow list only, and not adding any?

**Reply by rijkvanzanten** [reply]
That too of course, though that'd still have the system ones available 🤔

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
